File: cortex_web/apps/web/trainer/reward.ts
// One-step expected-reward scorer (G3 TS port of trainer/reward._expected_reward).
// The policy's bias-correction term wraps this. Deterministic; tolerance-parity
// gated. `learn` is always 1 here (the p_static mixture is an opt-in Python
// ablation, default off, and is not ported).
import { normCdf } from '../engine/mathfns';
import { LAPSE_RATE, SKILL_MODE_MULTIPLIER } from './conventions';
import type { FilterParams } from './filter';

export interface RewardWeights {
  betaT: number;
  betaSigma: number;
  betaR: number;
}

export interface CloudView {
  theta: number[];
  ell: number[];
  w: number[];
}

// Q(s): weighted expected one-step reward over the particle cloud, using the
// filter's OWN assumed dynamics (`params`). Soft-rule / y-independent form.
// s, yStar: (n,) candidate arrays → (n,) Q. sSd: scalar or (n,).
export function expectedReward(
  cloud: CloudView, params: FilterParams, s: number[], yStar: number[],
  weights: RewardWeights,
  opts: { retBonus?: number[]; sSd?: number | number[]; barEll?: number | null } = {},
): number[] {
  const n = s.length;
  const N = cloud.theta.length;
  const sSdArr = Array.isArray(opts.sSd)
    ? opts.sSd
    : new Array(n).fill(opts.sSd ?? 0);
  const Q = new Array(n).fill(0);
  const logSigInf = Math.log(params.sigmaInf);
  for (let c = 0; c < n; c++) {
    const sc = s[c];
    const yc = yStar[c];
    const ssd = sSdArr[c];
    const rb = opts.retBonus ? opts.retBonus[c] : 0;
    let q = 0;
    for (let i = 0; i < N; i++) {
      const sigma = Math.exp(-cloud.ell[i]);
      const t = -cloud.theta[i];
      const el = Math.exp(cloud.ell[i]);
      let z = el * (sc + cloud.theta[i]);
      z = z / Math.sqrt(1 + (el * ssd) ** 2);
      const pYes = LAPSE_RATE + (1 - 2 * LAPSE_RATE) * normCdf(z);
      const delta = pYes - yc;
      const tNew = t + params.alphaT * delta;
      const d = Math.abs(sc - t) / sigma;
      const v = params.rho ** 2 + (ssd / sigma) ** 2;
      const wWeight = (params.rho / Math.sqrt(v))
        * Math.exp(-((d - SKILL_MODE_MULTIPLIER) ** 2) / (2 * v));
      const logSigNew = Math.log(sigma)
        - params.alphaSigma * wWeight * (Math.log(sigma) - logSigInf);
      let sigGain = sigma - Math.exp(logSigNew);
      if (opts.barEll != null && !(cloud.ell[i] < opts.barEll)) sigGain = 0;
      const R = weights.betaT * (Math.abs(t) - Math.abs(tNew))
        + weights.betaSigma * sigGain + weights.betaR * rb;
      q += R * cloud.w[i];
    }
    Q[c] = q;
  }
  return Q;
}

File: trainer/reward.py
"""One-step expected-reward scorer (G1 port of trainer_rd trainer_greedy's shared
`RewardWeights` + `_expected_reward`). The tier-2 policy's bias-correction and
expected-progress scorers are thin wrappers over this. Verbatim numerics port;
bit-parity gated by `tests/test_trainer_g1_policy.py`. (The full tier-1
`greedy_select` ablation is not shipped.)
"""
from __future__ import annotations

from dataclasses import dataclass

import numpy as np
from scipy.special import ndtr

from trainer.conventions import SKILL_MODE_MULTIPLIER, LAPSE_RATE


@dataclass
class RewardWeights:
    beta_t: float = 1.0
    beta_sigma: float = 1.0
    beta_r: float = 0.5


def _expected_reward(filt, s, y_star, weights: RewardWeights,
                     ret_bonus=0.0, s_sd=0.0, bar_ell=None):
    """Q(s): weighted expected one-step reward over the particle cloud, using the
    filter's OWN assumed dynamics (the transition means) to form θ̄'. Soft rule ⇒
    y-independent. Accepts scalar s (returns float) or an (n,) array of candidate
    signals paired with (n,) y_star (returns (n,) Q) — the array form broadcasts
    the whole bank against the cloud in one shot.

    s_sd (F20): stimulus uncertainty; the prediction-error term uses the
    attenuated P(yes), the skill-weight term the noise-smeared expected weight.
    bar_ell (opt-in): credit σ-progress only for particles still below the bar."""
    s = np.atleast_1d(np.asarray(s, dtype=np.float64))[:, None]      # (n,1)
    y_star = np.atleast_1d(np.asarray(y_star, dtype=np.float64))[:, None]
    ret_bonus = np.atleast_1d(np.asarray(ret_bonus, dtype=np.float64))[:, None]
    s_sd = np.atleast_1d(np.asarray(s_sd, dtype=np.float64))
    s_sd = (np.zeros_like(s) + s_sd[:, None]) if s_sd.size > 1 else \
        np.full_like(s, float(s_sd[0]))
    p = filt.p
    sigma = np.exp(-filt.ell)[None, :]                               # (1,N)
    t = -filt.theta[None, :]
    el = np.exp(filt.ell)[None, :]
    z = el * (s + filt.theta[None, :])
    z = z / np.sqrt(1.0 + (el * s_sd) ** 2)            # attenuation (F20)
    p_yes = LAPSE_RATE + (1.0 - 2.0 * LAPSE_RATE) * ndtr(z)
    delta = p_yes - y_star                              # soft prediction error
    learn = filt.learn[None, :]                        # p_static mass ⇒ no reward
    t_new = t + learn * p.alpha_t * delta
    d = np.abs(s - t) / sigma
    v = p.rho ** 2 + (s_sd / sigma) ** 2               # smeared weight (F20)
    w_weight = (p.rho / np.sqrt(v)
                * np.exp(-((d - SKILL_MODE_MULTIPLIER) ** 2) / (2.0 * v)))
    log_sig_new = (np.log(sigma)
                   - learn * p.alpha_sigma * w_weight
                   * (np.log(sigma) - np.log(p.sigma_inf)))
    sigma_new = np.exp(log_sig_new)
    sig_gain = sigma - sigma_new
    if bar_ell is not None:                 # bar-referenced credit (F85)
        sig_gain = sig_gain * (filt.ell[None, :] < float(bar_ell))
    R = (weights.beta_t * (np.abs(t) - np.abs(t_new))
         + weights.beta_sigma * sig_gain
         + weights.beta_r * ret_bonus)
    Q = (R * filt.w[None, :]).sum(axis=1)
    return float(Q[0]) if Q.shape[0] == 1 else Q
